src/go/k8s/internal/client/factory.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/redpanda-data/helm-charts/pkg/redpanda"
+	redpandav1alpha1 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha1"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrEmptyBrokerList             = errors.New("empty broker list")
+	ErrEmptyURLList                = errors.New("empty url list")
+	ErrInvalidKafkaClientObject    = errors.New("cannot initialize Kafka API client from given object")
+	ErrInvalidRedpandaClientObject = errors.New("cannot initialize Redpanda Admin API client from given object")
+	ErrUnsupportedSASLMechanism    = errors.New("unsupported SASL mechanism")
+)
+
+// ClientFactory is responsible for creating both high-level and low-level clients used in our
+// controllers.
+//
+// Calling its `Kafka*` methods will initialize a low-level kgo.Client instance
+// based on the connection parameters contained within the corresponding CRD struct passed in
+// at method invocation.
+//
+// Calling its `RedpandaAdmin*` methods will initialize a low-level rpadmin.AdminAPI instance
+// based on the connection parameters contained within the corresponding CRD struct passed in
+// at method invocation.
+//
+// Additionally, high-level clients that act as functional wrappers to the low-level
+// clients to ease high-level operations (i.e. syncing all primitives for a v1alpha2.User) may
+// be initialized through calling the appropriate `*Client` methods.
+type ClientFactory interface {
+	// WithDialer causes the underlying connections to use a different dialer implementation.
+	WithDialer(dialer redpanda.DialContextFunc) ClientFactory
+	// WithLogger specifies a logger to be passed to the underlying clients.
+	WithLogger(logger logr.Logger) ClientFactory
+
+	// KafkaClient initializes a kgo.Client based on the spec of the passed in struct.
+	// The struct *must* implement either the v1alpha2.KafkaConnectedObject interface of the v1alpha2.ClusterReferencingObject
+	// interface to properly initialize.
+	KafkaClient(ctx context.Context, object client.Object, opts ...kgo.Opt) (*kgo.Client, error)
+
+	// RedpandaAdminClient initializes a rpadmin.AdminAPI client based on the spec of the passed in struct.
+	// The struct *must* implement either the v1alpha2.AdminConnectedObject interface of the v1alpha2.ClusterReferencingObject
+	// interface to properly initialize.
+	RedpandaAdminClient(ctx context.Context, object client.Object) (*rpadmin.AdminAPI, error)
+}
+
+type clientFactory struct {
+	client.Client
+	logger logr.Logger
+	config *rest.Config
+	dialer redpanda.DialContextFunc
+}
+
+var _ ClientFactory = (*clientFactory)(nil)
+
+func NewClientFactory(config *rest.Config) (ClientFactory, error) {
```

Comment:
I've had better experiences with constructing a single `scheme` in main.go that gets plumbed down everywhere rather than building schemes on the fly. Eventually we're going to have to update the scheme and tracking down all the schemes is super easy to miss because they only generate runtime errors :/

In this case it may be easier to accept a pre-built `client.Client`. As a bonus, you'd eliminate the need to return an error here.

Reply:
nit: When possible return the struct itself rather than an interface. This is particularly useful in tests as you can add helpers that aren't otherwise exposed.